Create rule S6424 (#964)
SonarSource/rspec: rules/S6424/csharp/rule.adoc

* Entity interfaces must only define methods.
* Entity interfaces must not contain generic parameters.
* Entity interface methods must not have more than one parameter.
* Entity interface methods must return void, Task, or Task<T>.

If any of these rules are violated, an `InvalidOperationException` is thrown at runtime when the interface is used as a type argument to `SignalEntity` or `CreateProxy`. The exception message explains which rule was broken.

This rule raises an issue in case any of the restrictions above is not respected.

// If you want to factorize the description uncomment the following line and create the file.
//include::../description.adoc[]

== Noncompliant Code Example

[source,csharp]
----
namespace Foo // Noncompliant, must be defined in the same assembly as the entity class that implements it
{
    public interface ICounter<T> // Noncompliant, interfaces cannot contain generic parameters
    {
        string Name { get; set; } // Noncompliant, interface must only define methods
        void Add(int amount, int second parameter); // Noncompliant, methods must not have more than one parameter

>         int Reset(); // Noncompliant, methods must return void, Task, or Task<T>

Reviewer: ```suggestion
        int Get(); // Noncompliant, methods must return void, Task, or Task<T>
```

(below you use Get for Task<int>)

    }
}

namespace Bar
{
    public class Counter : ICounter
    {
        // do stuff
    }
}
----

== Compliant Solution

[source,csharp]
----
namespace Bar
{
    public interface ICounter
    {

>         void Add(int amount);
>         Task Reset();
>         Task<int> Get();
>         void Delete();

Reviewer: ```suggestion
        void Add(int amount);
        Task<int> Get();
```

IMO it's ok to just show the fixed method

> The recommended way to use for accessing Azure Durable Entities is Interfaces via generated proxy objects.

Reviewer: ```suggestion
The recommended way to access Azure Durable Entities is Interfaces via generated proxy objects.
```

(please use grammarly)

Reviewer 2: I use it (all the rspecs I've opened PR for so far, have been scanned with grammarly), but it did not give me any correction there.
Still, this makes sense, I'll update it.

The following restrictions, during interface design, are enforced:

* Entity interfaces must be defined in the same assembly as the entity class.
* Entity interfaces must only define methods.
* Entity interfaces must not contain generic parameters.
* Entity interface methods must not have more than one parameter.
* Entity interface methods must return void, Task, or Task<T>.

> If any of these rules are violated, an `InvalidOperationException` is thrown at runtime when the interface is used as a type argument to `SignalEntity` or `CreateProxy`. The exception message explains which rule was broken.

Reviewer: ```suggestion
If any of these rules are violated, an `InvalidOperationException` is thrown at runtime when the interface is used as a type argument to `SignalEntity` or `CreateProxy.` The exception message explains which rule was broken.
```

(American English punctuation - I don't understand why, but it's the way it is)

    {
        string Name { get; set; } // Noncompliant, interface must only define methods
        void Add(int amount, int second parameter); // Noncompliant, methods must not have more than one parameter
        int Get(); // Noncompliant, methods must return void, Task, or Task<T>
    }
}

namespace Bar
{
    public class Counter : ICounter
    {
        // do stuff
    }
}
----

== Compliant Solution

[source,csharp]
----
namespace Bar
{
    public interface ICounter
    {
        void Add(int amount);
        Task<int> Get();
    }
}

namespace Bar
{
    public class Counter : ICounter
    {
        // do stuff
    }

    public static class AddToCounterFromQueue
    {
        // Azure Function Example using the Counter Durable Entity
        [FunctionName("AddToCounterFromQueue")]
        public static Task Run(
            [QueueTrigger("durable-function-trigger")] string input,
            [DurableClient] IDurableEntityClient client)
        {
            var entityId = new EntityId("Counter", "myCounter");
            int amount = int.Parse(input);
            return client.SignalEntityAsync<ICounter>(entityId, proxy => proxy.Add(amount));

>         }

Reviewer: this should also be present in the Noncompliant example, right?

Reviewer 2: Yes, I added the same Function.
